helix-core/src/selection.rs

```diff
@@ -335,6 +335,21 @@ impl Selection {
         self.ranges[self.primary_index]
     }
 
+    #[inline]
+    #[must_use]
+    pub fn remove_primary(&self) -> Self {
+        assert!(self.len() > 1);
+
+        let primary = self.ranges[self.primary_index];
+        let ranges = self.iter().filter(|r| **r != primary).cloned().collect();
+        let primary_index = if self.primary_index == self.len() - 1 {
+            0
+        } else {
+            self.primary_index
+        };
```

Comment:
```suggestion
        let mut ranges = self.ranges.clone();
        ranges.remove(self.primary_index);
        let primary_index = self.primary_index.saturating_sub(1);
```

The (subtle) difference is that the primary selection will move the other way when it's deleted since we subtract here.